src/main/java/com/iota/iri/service/milestone/impl/MilestoneSolidifierImpl.java

```diff
@@ -1,339 +1,666 @@
 
-import com.iota.iri.service.validation.TransactionSolidifier;
+import com.iota.iri.conf.SolidificationConfig;
+import com.iota.iri.controllers.AddressViewModel;
+import com.iota.iri.controllers.MilestoneViewModel;
+import com.iota.iri.controllers.TransactionViewModel;
 import com.iota.iri.model.Hash;
+import com.iota.iri.network.TransactionRequester;
+import com.iota.iri.service.ledger.LedgerService;
+import com.iota.iri.service.milestone.MilestoneException;
+import com.iota.iri.service.milestone.MilestoneService;
 import com.iota.iri.service.milestone.MilestoneSolidifier;
+import com.iota.iri.service.milestone.MilestoneValidity;
+import com.iota.iri.service.snapshot.Snapshot;
 import com.iota.iri.service.snapshot.SnapshotProvider;
+import com.iota.iri.service.validation.TransactionSolidifier;
+import com.iota.iri.storage.Tangle;
+import com.iota.iri.utils.ASCIIProgressBar;
 import com.iota.iri.utils.log.interval.IntervalLogger;
 import com.iota.iri.utils.thread.DedicatedScheduledExecutorService;
 import com.iota.iri.utils.thread.SilentScheduledExecutorService;
 
 import java.util.*;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.TimeUnit;
-
-/**
- * <p>
- * This class implements the basic contract of the {@link MilestoneSolidifier} interface.
- * </p>
- * <p>
- * It manages a map of unsolid milestones to collect all milestones that have to be solidified. It then periodically
- * issues checkSolidity calls on the earliest milestones to solidify them.
- * </p>
- * <p>
- * To save resources and make the call a little bit more efficient, we cache the earliest milestones in a separate map,
- * so the relatively expensive task of having to search for the next earliest milestone in the pool only has to be
- * performed after a milestone has become solid or irrelevant for our node.
- * </p>
- */
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+
 public class MilestoneSolidifierImpl implements MilestoneSolidifier {
-    /**
-     * Defines the amount of milestones that we "simultaneously" try to solidify in one pass.
-     */
-    private static final int SOLIDIFICATION_QUEUE_SIZE = 20;
+    private static final IntervalLogger log = new IntervalLogger(MilestoneSolidifierImpl.class);
+    // Max size fo the solidification queue
+    private static final int MAX_SIZE = 10;
 
-    /**
-     * Defines the interval in which solidity checks are issued (in milliseconds).
-     */
-    private static final int SOLIDIFICATION_INTERVAL = 100;
+    private Map<Hash, Integer> unsolidMilestones = new ConcurrentHashMap<>();
+    private Map<Hash, Integer> solidificationQueue = new ConcurrentHashMap<>();
+    private Map<Integer, Hash> seenMilestones = new ConcurrentHashMap<>();
+
+    private Map.Entry<Hash, Integer> oldestMilestoneInQueue = null;
+
+    private int latestMilestoneIndex;
+    private Hash latestMilestoneHash;
+    private int latestSolidMilestone;
```

Comment:
adding all those state fields is usually a red flag...
In a multithreaded environment you want to avoid this as much as possible

Reply:
To help alleviate some of the concern i've switched the logic to now use atomic integers and atomic references to try and shield some of the possible issues that could surface.